pkg/policy/policy.go

```diff
@@ -26,6 +27,18 @@ import (
 	"github.com/op/go-logging"
 )
 
+// Completion is an interface for an individual asynchronous success/failure
+// notification.
+type Completion interface {
+	Completed(bool)
+}
+
+// CompletionContainer is an interface for pooling asynchronous success/failure
+// notifications.
+type CompletionContainer interface {
+	AddCompletion() (Completion, time.Duration)
```

Comment:
Instead of explicitly using a timeout, why not pass a `context.Context` to every calls that currently takes a `CompletionContainer`? Using a context will allow your streamctrl logic to support both explicit cancelations from callers and timeouts.
Instead of directly dealing with timeouts, you'd only need to check whether the context is still valid with `ctx.Err() == nil` to determine whether to continue waiting for an ACK.

Reply:
Also, all functions that currently take a `CompletionContainer` as a parameter only either pass it down the call stack, or call `AddCompletion` only once. I'd directly pass a `Completion` parameter instead of a `CompletionContainer`.

It may also be worth trying to implement the Future pattern, because that's really what you try to do here. A future may simply be modeled with a `chan bool` here. The callee would write at most one value into the channel and close it, once the version is ACKed/NACKed by the proxy.
The functions' interfaces would then look like:
```
func (s *LDSServer) updateListener(ctx context.Context, ..., out chan<- bool) {
    ...
}
```

If you wanted to define a fancy type to wrap the channel details, you could pass a `Completion` to do the job, and implement it so:
```
type Completion func(success bool)

func NewCompletion(ctx Context, out chan<- bool) Completion {
    return func(success bool) {
        select {
        case <-ctx.Done():
            // Ignore.
        case out <- success:
        }
        close(out)
    }
}

func (s *LDSServer) updateListener(ctx context.Context, ..., completion Completion) {
    ...
}
```

Reply:
Now I think you'll always want to combine a completion with a context, so it may be better to define the interface so:
```
type Completion interface {
    Context() context.Context
    Completed(success bool)
}
```

And just pass only that around, e.g.:
```
func (e *Envoy) UpdateListener(name string, l7rules policy.L7DataMap, completion policy.Completion) {
    ...
}
```

Reply:
Part of my design was to both:
- Allow passing CompletionContainer as nil to support callers who may not be interested in the completion signal
- Allow implementations that are always synchronous to ignore the CompletionContainer

Furthermore, my implementation allows:
- the implementation of a Completion interface package a single WaitGroup
- and share that same instance for multiple Completions in the same CompletionsContainer

This allows for a potentially large number of completions to be waited for with a single WaitGroup.Wait().

To do the same with a single channel without blocking writers I'd have to have a buffered channel for some max number of completions?

Reply:
I'll have to look at the timeout.. Maybe it would be better to have an overall daedline for the whole policy update operartion rather than individual timeouts on each completion. If so the Context() could be a member in the CompletionContainer interface. I guess context.Context must be passed by reference for a cancellation to pass through?